### ISSUE TITLE
disabling css in umatrix prevents some (Print Edit WE) firefox extensions from working

### ISSUE THREAD

**Issue by AsocialSurvival** [uBlockOrigin/uMatrix-issues#115]
### Prerequisites

- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
    - [x] I also searched the existing issues at <https://github.com/gorhill/uMatrix/issues>
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uMatrix](https://old.reddit.com/r/uMatrix/)
- I tried to reproduce the issue when...
    - [x] uMatrix is the only extension
    - [x] uMatrix with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uMatrix
- [x] I checked the [documentation](https://github.com/gorhill/uMatrix/wiki) to understand that the issue I report is not a normal behavior
- [x] I used the logger to rule out that the issue is caused by my ruleset

### Description
umatrix prevents firefox extension "Print Edit WE" from working (displaying it's overlay menu (below address or bookmarks bar, or on top of page)) when css files are disabled in any case (from specific or all sites), i think it may prevent much other extensions from working because they may be based on css, because they probably inject or modify css and other elements of current page

### A specific URL where the issue occurs
https://www.ghacks.net/2018/03/10/firefox-60-ships-with-windows-group-policy-support/

### Steps to Reproduce

1. Install Print Edit WE and your extension (umatrix)
2. Go to https://www.ghacks.net/2018/03/10/firefox-60-ships-with-windows-group-policy-support/
3. Disable all css, or only ones from www.ghacks.net (i needed this because ghacks can't be printed normally, but only shows part of page, probably because of css, like many other problematic pages)
4. Press Print Edit WE button that appears on toolbar (next to umatrix icon), you will see nothing happens but page is in edit mode in their extension, similar to print preview mode, and that pressing their button again will lead into other mode, but their ribbon (bar) that has option to close that mode will never appear, so the only way to close their mode is to refresh page, (and select Leave in firefox dialog box if asked). Or nothing happens.

### Ruleset
permenent rules:
https-strict: behind-the-scene false
matrix-off: about-scheme true
matrix-off: behind-the-scene true
matrix-off: chrome-extension-scheme true
matrix-off: chrome-scheme true
matrix-off: moz-extension-scheme true
matrix-off: opera-scheme true
matrix-off: vivaldi-scheme true
matrix-off: wyciwyg-scheme true
noscript-spoof: * true
noscript-spoof: ghacks.net false
referrer-spoof: * true
referrer-spoof: behind-the-scene false
referrer-spoof: ghacks.net false
* * * allow
* googlesyndication.com * inherit
* pagead2.googlesyndication.com * inherit
ghacks.net googlesyndication.com * inherit
ghacks.net pagead2.googlesyndication.com * inherit

temporary rules:
https-strict: behind-the-scene false
matrix-off: about-scheme true
matrix-off: behind-the-scene true
matrix-off: chrome-extension-scheme true
matrix-off: chrome-scheme true
matrix-off: moz-extension-scheme true
matrix-off: opera-scheme true
matrix-off: vivaldi-scheme true
matrix-off: wyciwyg-scheme true
noscript-spoof: * true
noscript-spoof: ghacks.net false
referrer-spoof: * true
referrer-spoof: behind-the-scene false
referrer-spoof: ghacks.net false
* * * allow
* googlesyndication.com * inherit
* pagead2.googlesyndication.com * inherit
ghacks.net googlesyndication.com * inherit
ghacks.net pagead2.googlesyndication.com * inherit
ghacks.net www.ghacks.net css block

### Supporting evidence

here are two cases in 4 pictures, when css is disabled and when it is enabled, meaning when there is problem and when not, of course page has to be refreshed after enabling or disabling css in your extension
![2](https://user-images.githubusercontent.com/40727207/53204584-4411f800-362c-11e9-905c-c0218b2a8310.jpg)
![1](https://user-images.githubusercontent.com/40727207/53204585-4411f800-362c-11e9-820d-817af18527cb.jpg)
![4](https://user-images.githubusercontent.com/40727207/53204805-d4503d00-362c-11e9-901e-11f5c80594f4.jpg)
![3](https://user-images.githubusercontent.com/40727207/53204806-d4e8d380-362c-11e9-8e9d-8487b8ebf16d.jpg)

### Your environment

* uMatrix version: 1.18.4
* Browser Name and version: Mozilla Firefox 65.0.1 (32-bit)
* Operating System and version: Windows 7 professional 32bit with sp1 + convenient rollup update

**Comment by gorhill** [comment]
That is not a uMatrix issue, this is a Print Edit WE issue: that extension will malfunction without uMatrix on any site which has a CSP directive forbidding the use of inline styles. The author of that extension needs to fox it this way: <https://github.com/gorhill/uBlock/commit/ed492bc48578f42a0f6d6893e3ef43b168ea57c2#diff-a65f6e6abc235a14e93548d8bf53b81eR1508> -- re-creating the inline styles such that Firefox will bypass the page's CSP `style-src`.

**Comment by AsocialSurvival** [comment]
OK, I will report to them. I found solution in the meantime, in fact preferable and permanent then using any other. Remove (do not use) extensions Save Page WE and umatrix. Then using only ublock block all css from that webpage/website. Then using ublock remove unnecessary elements, and finally print! ublock is already good enough for this.

**Comment by gorhill** [comment]
Here is an example of a site where it does not work, no other extension installed:
https://www.mnot.net/blog/2018/08/14/trust_australia

**Comment by AsocialSurvival** [comment]
Wow, you're right, Luckily ublock alone is enough to help me deal with my personal issue:
- inability to print some pages (they only show part of text or sometimes comments only or part of images) because of css, disabling css reveals and makes all content printable although changes appearance of page into worse, plus remove all unnecessary content via element blocker, i could not do it previously when saved web pages as mht or htm, but now when print to pdf it is ok, plus much smaller size.

**Comment by AsocialSurvival** [comment]
The author said:

There several reasons why a page will not print correctly, for example, :
  • Flex styling.
  • Very long tables.
  • body or div elements with float or non-relative position.
  • Poorly designed print media style sheets.
  • Bugs in Firefox’s print preview functionality.
 
Print Edit WE has several features that can help work around these issues:
  • Web Style – to avoid using the print media style sheets.
  • Fix Page Breaks – to avoid problems with body or div elements.
  • Remove Flex Styles – to avoid problems with flex styling.
  • Format Panel – to insert or avoid page breaks.

Print Edit WE relies on CSS and in-line styles in order to perform its editing and formatting functions. So blocking CSS or in-line styles using uMatrix or CSP ‘style-src’ will prevent Print Edit WE performing its main functions.  It will also prevent the Print Edit WE toolbar being displayed.
 
The solution proposed by Raymond Hill (gorhill) would allow the toolbar to be displayed, but would not allow the main editing and formatting functions.
 
It might be possible to avoid using in-line styles for editing and formatting by adding rules to existing external stylesheets or by injecting stylesheets, but this would probably require adding identifiers to most elements in the page to allow targeting of the rules.  This feels clunky and doesn’t feel like a good solution.

**Comment by AsocialSurvival** [comment]
Ahhh, finally found a solution. Looks like Print Edit WE can help me! I did not know how to use it. All I had to do was to press on their icon, then tools button, and fix page breaks! No need for blocking css, or other extreme measures. Now in print preview all is already good. Tested on both websites! While wikihow started showing images when only pressed on use "Web Style" I feel so stupid for not figuring this out immediately! So yeah, this solved all my problems. I'll consider this as solved, but if you have better solution (i doubt), then post it. Many people complained for this problem, and Print Edit WE is the real solution! Specifically these 3 options (buttons) I used. Maybe only adblockers like ublock are better for permanent change of pages look (excess buttons, ads...). I also tried saving as mht ot html or html via save page we (which is actually like mht), but removing buttons does not make changes to such files, but makes changes to printing. Also mht take much more space, the only advantage being that they keep urls.
